File: app/src/main/java/com/example/schooldiary/ui/registration/createAccount/CreateUserFragment.java
package com.example.schooldiary.ui.registration.createAccount;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;

import androidx.annotation.VisibleForTesting;
import androidx.fragment.app.Fragment;

import android.text.TextUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.example.schooldiary.R;
import com.example.schooldiary.ui.MainActivity;
import com.google.android.material.textfield.TextInputEditText;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;



public class CreateUserFragment extends Fragment {

    private static final String TAG = "EmailPassword";

    private TextView mStatusTextView;
    private TextView mDetailTextView;
    private TextInputEditText mEmailField;
    private TextInputEditText mPasswordField;

    @VisibleForTesting
    public ProgressBar mProgressBar;

    // [START declare_auth]
    private FirebaseAuth mAuth;
    // [END declare_auth]

    @Override
    public void onStart() {
        super.onStart();
        // Check if user is signed in (non-null) and update UI accordingly.
        FirebaseUser currentUser = mAuth.getCurrentUser();
        updateUI(currentUser);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_create_user, container, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        // Views
        mStatusTextView = getActivity().findViewById(R.id.registration_email_password_status);
        mDetailTextView = getActivity().findViewById(R.id.registration_detail_status);
        mEmailField = getActivity().findViewById(R.id.registration_text_input_email);
        mPasswordField = getActivity().findViewById(R.id.registration_text_input_password);
        setProgressBar(R.id.registration_progress_bar);
        // hide here, otherwise bar will spin
        hideProgressBar();


        // [START initialize_auth]
        // Initialize Firebase Auth
        mAuth = FirebaseAuth.getInstance();
        // [END initialize_auth]


        Button signUp = view.findViewById(R.id.sign_up_btn_2);
        signUp.setOnClickListener(v -> {
            createAccount(view, mEmailField.getText().toString(), mPasswordField.getText().toString());

        });
    }

    private void createAccount(View view, String email, String password) {
        Log.d(TAG, "createAccount:" + email);
        if (!validateForm()) {
            return;
        }

        showProgressBar();

        // [START create_user_with_email]
        mAuth.createUserWithEmailAndPassword(email, password)
                .addOnCompleteListener(task -> {
                        if (task.isSuccessful()) {
                            // Sign in success, update UI with the signed-in user's information
                            Log.d(TAG, "createUserWithEmail:success");
                            FirebaseUser user = mAuth.getCurrentUser();
                            updateUI(user);
                        } else {
                            // If sign in fails, display a message to the user.
                            Log.w(TAG, "createUserWithEmail:failure", task.getException());
                            Toast.makeText(view.getContext(), "Registration failed.",
                                    Toast.LENGTH_SHORT).show();
                        }

                        // [START_EXCLUDE]
                        // when auth is fail hide on time, but when auth is success hide much earlier than a new activity begins
                        hideProgressBar();
                        // [END_EXCLUDE]
                });
        // [END create_user_with_email]
    }

    public void setProgressBar(int resId) {
        mProgressBar = getActivity().findViewById(resId);
    }

    public void showProgressBar() {
        if (mProgressBar != null) {
            mProgressBar.setVisibility(View.VISIBLE);
        }
    }

    public void hideProgressBar() {
        if (mProgressBar != null) {
            mProgressBar.setVisibility(View.INVISIBLE);
        }
    }

    private boolean validateForm() {
        boolean valid = true;

        String email = mEmailField.getText().toString();
        if (TextUtils.isEmpty(email)) {
            mEmailField.setError("Required.");
            valid = false;
        } else {
            mEmailField.setError(null);
        }

        String password = mPasswordField.getText().toString();
        if (TextUtils.isEmpty(password)) {
            mPasswordField.setError("Required.");
            valid = false;
        } else {
            mPasswordField.setError(null);
        }

        return valid;
    }

    private void updateUI(FirebaseUser user) {
        if (user != null) {
            mStatusTextView.setText(getString(R.string.emailpassword_status_fmt,
                    user.getEmail(), user.isEmailVerified()));
            mDetailTextView.setText(getString(R.string.firebase_status_fmt, user.getUid()));

            Activity activity = getActivity();
            Intent intent = new Intent(activity, MainActivity.class);
            startActivity(intent);
            activity.finish();

        }
    }
}

File: app/src/main/java/com/example/schooldiary/ui/registration/createAccount/CreateUserFragment.kt
package com.example.schooldiary.ui.registration.createAccount

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.annotation.VisibleForTesting
import androidx.fragment.app.Fragment
import com.example.schooldiary.R
import com.example.schooldiary.databinding.FragmentCreateUserBinding
import com.example.schooldiary.ui.MainActivity
import com.google.android.gms.tasks.Task
import com.google.android.material.textfield.TextInputEditText
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

class CreateUserFragment : Fragment() {
    private var _binding: FragmentCreateUserBinding? = null
    private val binding
        get() = _binding!!

    private lateinit var statusTextView: TextView
    private lateinit var detailTextView: TextView
    private lateinit var emailField: TextInputEditText
    private lateinit var passwordField: TextInputEditText

    @VisibleForTesting
    private lateinit var progressBar: ProgressBar

    private lateinit var auth: FirebaseAuth

    override fun onStart() {
        super.onStart()
        // Check if user is signed in (non-null) and update UI accordingly.
        val currentUser = auth.currentUser
        updateUI(currentUser)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_create_user, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {

        _binding = FragmentCreateUserBinding.bind(view)

        statusTextView = binding.registrationEmailPasswordStatus
        detailTextView = binding.registrationDetailStatus
        emailField = binding.registrationTextInputEmail
        passwordField = binding.registrationTextInputPassword
        progressBar = binding.registrationProgressBar
        // hide here, otherwise bar will spin
        hideProgressBar()


        auth = FirebaseAuth.getInstance()

        val signUp = binding.signUpBtn2
        signUp.setOnClickListener {
            createAccount(view, emailField.text.toString(), passwordField.text.toString())
        }
    }

    private fun createAccount(view: View, email: String, password: String) {
        Log.d(TAG, "createAccount:$email")
        if (!validateForm()) {
            return
        }
        showProgressBar()

        GlobalScope.launch(Dispatchers.IO) {
            auth.createUserWithEmailAndPassword(email, password)
                    .addOnCompleteListener { task: Task<AuthResult?> ->
                        if (task.isSuccessful) {
                            // Sign in success, update UI with the signed-in user's information
                            Log.d(TAG, "createUserWithEmail:success")
                            val user = auth.currentUser
                            GlobalScope.launch(Dispatchers.Main) {
                                updateUI(user)
                            }
                        } else {
                            // If sign in fails, display a message to the user.
                            Log.w(TAG, "createUserWithEmail:failure", task.exception)
                            GlobalScope.launch(Dispatchers.Main) {
                                Toast.makeText(view.context, "Registration failed.",
                                        Toast.LENGTH_SHORT).show()
                            }
                        }

                        // when auth is fail hide on time, but when auth is success hide much earlier than a new activity begins

//                        GlobalScope.launch(Dispatchers.Main) {
//                            hideProgressBar()
//                        }
                    }
        }
    }

    private fun createUser(view: View, email: String, password: String) {

    }

    private fun showProgressBar() {
        progressBar.visibility = View.VISIBLE
    }

    private fun hideProgressBar() {
        progressBar.visibility = View.INVISIBLE
    }

    private fun validateForm(): Boolean {
        var valid = true
        val email = emailField.text.toString()
        if (TextUtils.isEmpty(email)) {
            emailField.error = "Required."
            valid = false
        } else {
            emailField.error = null
        }
        val password = passwordField.text.toString()
        if (TextUtils.isEmpty(password)) {
            passwordField.error = "Required."
            valid = false
        } else {
            passwordField.error = null
        }
        return valid
    }

    private fun updateUI(user: FirebaseUser?) {
        if (user != null) {
            statusTextView.text = getString(R.string.emailpassword_status_fmt,
                    user.email, user.isEmailVerified)
            detailTextView.text = getString(R.string.firebase_status_fmt, user.uid)
            val activity: Activity? = activity
            val intent = Intent(activity, MainActivity::class.java)
            startActivity(intent)
            activity!!.finish()
        }
    }

    companion object {
        private const val TAG = "EmailPassword"
    }
}